### PR TITLE
Fix network api accumulating connection handlers and not firing some disconnect events

### PR DESCRIPTION
**Pull request by Player3324**
The existing networking-api implementation doesn't maintain GlobalReceiverRegistry.trackedAddons properly, accumulating stale Network(Handler)Addon instances. It also doesn't fire the disconnect event reliably, in particular the clientside disconnect event doesn't happen when leaving a SP game normally through the menu.

I fixed the issue by clearing LoginNetworkAddon instances when client/server transition from LOGIN to PLAY and by hooking the Netty-driven ClientConnection.channelInactive to capture all disconnects eventually. I left the regular disconnect hooks in to avoid potentially introducing extra delays in case the vanilla code doesn't close the connection immediately besides other potential regressions.

This should fix https://github.com/FabricMC/fabric/issues/1439

### CODE REVIEW

**Comment by Ampflower** [comment]
Can confirm that this does fix #1439. Reloaded the world several times and saw that the minimum memory after GC was relatively the same each time. Also took a quick look at the heap dump via VisualVM and saw only one instance of server, which was the one that's currently live.

**Review by modmuss50 — APPROVED** [review]
LGTM.

This seems worthy of backporting to 1.14 and 1.15

File: fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/AbstractNetworkAddon.java [hunk 1]
```diff
@@ -44,6 +45,7 @@
 	// Sync map should be fine as there is little read write competition
 	// All access to this map is guarded by the lock
 	private final Map<Identifier, H> handlers = new HashMap<>();
+	private AtomicBoolean disconnected = new AtomicBoolean(); // blocks redundant disconnect notifications
```
Comment (Technici4n) [inline fabric-networking-api-v1/src/main/java/net/fabricmc/fabric/impl/networking/AbstractNetworkAddon.java]: `final` maybe?

**Review by Technici4n — APPROVED** [review]
LGTM, tested this on AOF3 and it seems to fix the issue in singleplayer, and multiplayer still works.

**Pull request merged**